godot/components/entities/behaviors/scripts/PatrolBehavior.gd

```diff
@@ -0,0 +1,29 @@
+extends EntityBehavior
+class_name PatrolBehavior
+
+
+@export var speed = 100.0
+@export var edge_detecting_x_distance = 8
+
+func apply_behavior(entity: Entity, delta: float):
+	# Add the gravity.
+	if not entity.is_on_floor():
+		entity.velocity.y += entity.gravity * delta
+		entity.velocity.y = minf(entity.velocity.y, entity.MAX_FALL_SPEED)
+		
+	if entity.is_on_wall() || _is_near_edge(entity):
+		scale.x = -1
+		entity._is_facing_right = !entity._is_facing_right
+		
+	entity.velocity.x = (1 if entity._is_facing_right else -1) * speed
+	
+func _is_near_edge(entity: Entity):
+	if edge_detecting_x_distance <= 0 || !entity.is_on_floor():
+		return false
+		
+	var space_state = get_world_2d().direct_space_state
+	var x_position = entity.global_position.x + (edge_detecting_x_distance if entity._is_facing_right else -edge_detecting_x_distance)
+	var edge_raycast = PhysicsRayQueryParameters2D.create(Vector2(x_position, entity.global_position.y - 2), Vector2(x_position, entity.global_position.y + 2))
+	var edge_result = space_state.intersect_ray(edge_raycast)
```

Comment:
would love an explainer of this at some point!

Reply:
Sure! The basic idea is that line 26 creates a slightly downward angled ray from just above the entity's feet the just below the entity's feet, angled toward where the entity is looking. If that day collides with ground (space_state), that means that the entity can walk forward since there is ground in front of it. Otherwise it needs to turn around.